### ISSUE TITLE
Pickle Crashes for Large Data

### ISSUE THREAD

**Issue by adam-grant-hendry** [pyvista/pyvista#3269]
### Describe the bug, what's wrong, and what you expected.

`pickle` crashes for large data.

### Steps to reproduce the bug.

```python
import pickle

import numpy as np
import pyvista as pv
from vtkmodules.vtkIOLegacy import vtkDataSetWriter


def pickle_vtk(mesh, filename):
    writer = vtkDataSetWriter()
    writer.SetInputDataObject(mesh)
    writer.SetWriteToOutputString(True)
    writer.SetFileTypeToBinary()
    writer.Write()
    to_serialize = writer.GetOutputString()

    with open(filename, 'wb') as handle:
        pickle.dump(to_serialize, handle, protocol=pickle.HIGHEST_PROTOCOL)

    return filename


if __name__ == '__main__':
    dims = (2154, 1500, 1167)

    volume = pv.UniformGrid(
        dims=(2154, 1500, 1167),
        spacing=(1, 1, 1),
        origin=(0, 0, 0),
    )

    volume.point_data['scalars'] = np.zeros(
        shape=(dims[0] * dims[1] * dims[2],), dtype=np.uint8
    )

    fname = 'filename.vtkpickle'
    pickle_vtk(volume, fname)  # Crashes
```

However, I have no problem pickling the numpy array by itself to file (this is 3.7 GB).

### System Information

```shell
--------------------------------------------------------------------------------
  Date: Sun Sep 04 11:59:36 2022 Pacific Daylight Time

                OS : Windows
            CPU(s) : 96
           Machine : AMD64
      Architecture : 64bit
               RAM : 190.7 GiB
       Environment : Python
        GPU Vendor : NVIDIA Corporation
      GPU Renderer : Quadro RTX 8000/PCIe/SSE2
       GPU Version : 4.5.0 NVIDIA 516.40

  Python 3.8.10 (tags/v3.8.10:3d8993a, May  3 2021, 11:48:03) [MSC v.1928 64
  bit (AMD64)]

           pyvista : 0.37.dev0
               vtk : 9.1.0
             numpy : 1.23.2
           imageio : 2.21.2
           appdirs : 1.4.4
            scooby : 0.5.12
        matplotlib : 3.5.3
         pyvistaqt : 0.9.0
           IPython : 8.4.0
              tqdm : 4.64.0
            meshio : 5.3.4
--------------------------------------------------------------------------------
```

**Comment by adam-grant-hendry** [comment]
@akaszynski `VtkErrorCatcher` doesn't output any results, so I can't debug from python. The writer is on the C++ side in a DLL.

**Comment by adeak** [comment]
Having taken a closer look at this issue (with more information than the other comments I've been pinged on), let me amend my previous comment:
1. I now see this is a Windows machine, so OOM killing doesn't enter the picture. I'm not familiar with Windows, but my vague understanding is that processes can still be terminated if it requests more memory than available.
2. I now also see that you have 190 GB of RAM. Is that all available? Or is the free memory considerably smaller?

**Comment by adam-grant-hendry** [comment]
> Is the script also killed if you leave out the call to pickle.dump?

Yes. When debugging, I find it crashes when it hits `writer.Write()` every time; it never gets to `pickle.dump`.

> I now also see that you have 190 GB of RAM. Is that all available? Or is the free memory considerably smaller?

On average, I tend to have at least 120 GB free available to use. This is an FEA and AI machine, so it is beefy.

![image](https://user-images.githubusercontent.com/59346180/188498779-a6ce1855-9156-4317-bc99-d284ea3d1e0c.png)

Here is a copy of my system info I generated in [PR #3170](https://github.com/pyvista/pyvista/pull/3170#issuecomment-1217191117)

<details>
<summary><b>System Info</b></summary>

```powershell
==================================System Info===================================
OS: Windows
Version: 10.0, Build 19042
Machine: AMD64
====================================CPU Info====================================
Processor: Intel64 Family 6 Model 85 Stepping 7, GenuineIntel
Physical Cores: 48
Total Cores: 96
=================================Virtual Memory=================================
Total: 190.65 GB
Available: 133.87 GB
==================================Swap Memory===================================
Total: 217.65 GB
Available: 153.52 GB
====================================GPU Info====================================
Name             Total Memory    Free Memory  
---------------  --------------  -------------
Quadro RTX 8000  46080.0 MB      45662.0 MB   
Quadro RTX 8000  46080.0 MB      43493.0 MB   
Quadro RTX 4000  8192.0 MB       6312.0 MB    
=====================================Python=====================================
Python: 3.8.10
======================================VTK=======================================
VTK: 9.1.0
```
</details>

**Comment by whophil** [comment]
@adam-grant-hendry does it also crash if you use `vtkDataSetWriter` to write to a file instead of a string?

**Comment by adeak** [comment]
I'm struggling to find what's going on in VTK, for what it's worth. [`vtkDataSetWriter`](https://vtk.org/doc/nightly/html/classvtkDataSetWriter.html) only seems to have a `WriteData()` method, but no `Write()`. [Its `WriteData()` method](https://gitlab.kitware.com/vtk/vtk/-/blob/master/IO/Legacy/vtkDataSetWriter.cxx#L35) dispatches to [a `vtkStructuredPointsWriter`](https://gitlab.kitware.com/vtk/vtk/-/blob/master/IO/Legacy/vtkDataSetWriter.cxx#L51-56), and eventually calls its [`Write()` method](https://gitlab.kitware.com/vtk/vtk/-/blob/master/IO/Legacy/vtkDataSetWriter.cxx#L97). But, again, I can't find this method [being defined](https://gitlab.kitware.com/vtk/vtk/-/blob/master/IO/Legacy/vtkStructuredGridWriter.cxx) nor [documented](https://vtk.org/doc/nightly/html/classvtkStructuredPointsWriter.html#details). Is this functionality documented? Am I missing something here?

The reason why I started to look at this is that I wanted to see what we're trying to pickle. For a small grid with `dim = (2, 2, 2)` the string we're pickling is
```py
>>> print(data)
# vtk DataFile Version 5.1
vtk output
BINARY
DATASET STRUCTURED_POINTS
DIMENSIONS 2 2 2
SPACING 1 1 1
ORIGIN 0 0 0
POINT_DATA 8
COLOR_SCALARS scalars 1

>>> data
'# vtk DataFile Version 5.1\nvtk output\nBINARY\nDATASET STRUCTURED_POINTS\nDIMENSIONS 2 2 2\nSPACING 1 1 1\nORIGIN 0 0 0\nPOINT_DATA 8\nCOLOR_SCALARS scalars 1\n'
```
For a moment I thought something was mangled, but it turns out VTK omits the scalars array if it's all zeros. Compare an array where every value is set to 1:
```py
>>> data
'# vtk DataFile Version 5.1\nvtk output\nBINARY\nDATASET STRUCTURED_POINTS\nDIMENSIONS 2 2 2\nSPACING 1 1 1\nORIGIN 0 0 0\nPOINT_DATA 8\nCOLOR_SCALARS scalars 1\n\x01\x01\x01\x01\x01\x01\x01\x01\n'
```
I'm actually half-sure that this might be a bug: if I set all the scalars to 1 except for the first one (which I set to 0) I get
```py
>>> data
'# vtk DataFile Version 5.1\nvtk output\nBINARY\nDATASET STRUCTURED_POINTS\nDIMENSIONS 2 2 2\nSPACING 1 1 1\nORIGIN 0 0 0\nPOINT_DATA 8\nCOLOR_SCALARS scalars 1\n'
```
This smells like zero bytes prematurely signalling the end of a C string...

Aaaanyway, what this all told me is that the size of the string in your snippet should be trivial. About as large as the ones I had above. And, correspondingly, you said you get the crash during the call to `writer.Write()`. So before the (short) string is built via `writer.GetOutputString()`.

If `writer.Write()` allocates space then it's still possible that my "trying to overallocate" theory is correct. But I can't find where `writer.Write()` is implemented, see the start of this comment. Input from more VTK-savvy people welcome.

**Comment by whophil** [comment]
> I'm struggling to find what's going on in VTK, for what it's worth. [`vtkDataSetWriter`](https://vtk.org/doc/nightly/html/classvtkDataSetWriter.html) only seems to have a `WriteData()` method, but no `Write()`

`vtkDataSetWriter.Write()` is inherited from `vtkWriter` (see https://vtk.org/doc/nightly/html/classvtkDataSetWriter-members.html)

**Comment by whophil** [comment]
@adam-grant-hendry with one modification, the snippet you posted runs fine on my machine which has much less memory than yours, producing a 3.5 GB file.

Change
```python
    to_serialize = writer.GetOutputString()
```

to 
```python
    to_serialize = writer.GetOutputStdString()
```

This method is used in the pickling code of PyVista, not `GetOutputString`.
https://github.com/pyvista/pyvista/blob/cf5b340201d68a0402e688d5ac40fe46aaeb7b99/pyvista/core/dataobject.py#L545

@adeak I think you'll find that your issue with all zero point data is a red herring, as it stems from the return of `GetOutputString()` having been truncated, for some reason, when there are zeros in the data.

**Comment by whophil** [comment]
@banesullivan I am also a bit confused by this particular snippet. I assume it was just written to debug why pickling a `DataObject` was failing, but after looking at your referenced thread, I see that may not be the case.

@adam-grant-hendry could you explain what you are trying to do?

Nearly all of the code you have written in the snippet is built into `DataObject` now, implemented by `DataObject.__setstate__` and `DataObject.__getstate__`. This means that `DataObject`s are automatically pickled and unpickled in case they are used in `multiprocessing`.

As @banesullivan mentioned, there is not really an use case for pickling the byte-serialization of a `DataObject` to a file as you are doing in your example.

**Comment by adam-grant-hendry** [comment]
> ...eventually calls its [Write() method](https://gitlab.kitware.com/vtk/vtk/-/blob/master/IO/Legacy/vtkDataSetWriter.cxx#L97). But, again, I can't find this method [being defined](https://gitlab.kitware.com/vtk/vtk/-/blob/master/IO/Legacy/vtkStructuredGridWriter.cxx) nor [documented](https://vtk.org/doc/nightly/html/classvtkStructuredPointsWriter.html#details). Is this functionality documented? Am I missing something here?

@whopil [got there before me](https://github.com/pyvista/pyvista/issues/3269#issuecomment-1237500285), but `vtkDataSetWriter` is a sublcass of `vtkWriter` and `Write()` is a public virtual method of `vtkWriter`. Since it's not pure virtual, this means it has a default implementation. The source is [here in IO/Core/vtkWriter.cxx](https://gitlab.kitware.com/vtk/vtk/-/blob/master/IO/Core/vtkWriter.cxx#L61). Hence, `vtkDataSetWriter` is calling this.

> This smells like zero bytes prematurely signaling the end of a C string...

I have an actual dataset. I used zeros as a demo. You could fill it with ones or whatever data you want and it will still crash. My dataset is a 3.7 GB micro-CT scan, but I cannot share it with you. It is proprietary.

> I don't know if this will fix the problems on your machine, which you state are happening when Write() is invoked (before GetOutputString()).

Unfortunately, you are correct. This will not solve the problem.

> You should just use `.save()` on the mesh to get a file created by a dedicated reader/writer for the mesh type at hand rather than the vtkDataSetWriter.

`save()` calls `writer.Write()`, which is where the problem is. This won't fix anything.

> It's tough for me to see any benefit/value from "pickling" to a file like this when in reality, you are just saving the contents of the VTK file writer to a file and calling it a "pickle"...I somewhat regret adding support for "pickling" like this

Pickling is useful when writing a GUI using `pyvistaqt`, or similar (see also [this user's comment](https://github.com/pyvista/pyvista-support/issues/59#issuecomment-945997117)). In a GUI, you typically have more than just a `VTK` dataset to save: you must also recreate some of the GUI internal structure or additional data. If a user opens a file associated with the GUI, they expect the GUI, `VTK` data, and everything else they made open up with it.

Ultimately, one could make a separate `xml` file specifically for the GUI and append the corresponding `VTK` data in it, which arguably might be the better long-term solution. However, for support of marshaling data, there is a benefit to pickling.

**Comment by adeak** [comment]
To focus on the issue at hand (as in pyvista issue), I'd like to recap:
1. [the snippet in the original comment here](https://github.com/pyvista/pyvista/issues/3269#issue-1361227150) crashes for @adam-grant-hendry on windows
2. it runs fine for @whophil on a system with much less memory ("runs fine" meaning `writer.Writer()` returns successfully).
3. since all of this is VTK, this is a potential VTK issue, and independent from pickling.

If we all agree, my recommendation would be to reduce the example to remove pyvista (by using a native VTK grid) and remove pickling (which is a red herring) and if the code still crashes for Adam and still doesn't crash for Phil, compare the two setups (OS, VTK version mainly) and open an issue with VTK. They will be able to tell whether what we see is expected (due to some memory management quirk) or a bug.

This will also affect PyVista through our pickling mechanism, but as long as we rely on these VTK writers we'll have to wait for such potential bugs to be fixed upstream.

**Comment by adam-grant-hendry** [comment]
@adeak I'm in agreement. I've modified the snippet to be just ``VTK 9.1`` without pickling, and it still crashes for me

```python
from vtkmodules.vtkCommonCore import vtkDoubleArray
from vtkmodules.vtkCommonDataModel import vtkImageData
from vtkmodules.vtkIOLegacy import vtkDataSetWriter

if __name__ == '__main__':
    points = vtkIntArray()
    points.SetName('points')
    points.SetNumberOfComponents(1)

    # 1-byte per int * 2154 * 1500 * 1167 ~= 3.77 GB
    points.SetNumberOfTuples(2154 * 1500 * 1167)

    volume = vtkImageData()
    volume.SetOrigin(0, 0, 0)
    volume.SetSpacing(1, 1, 1)
    volume.SetDimensions(2154, 1500, 1167)

    volume.GetPointData().SetScalars(points)

    writer = vtkDataSetWriter()
    writer.SetInputDataObject(volume)
    writer.SetWriteToOutputString(True)
    writer.SetFileTypeToBinary()
    writer.Write()
    to_serialize = writer.GetOutputString()

    print('Finished!')
    print(to_serialize)
```

Starting at 42.5 GB RAM

![image](https://user-images.githubusercontent.com/59346180/188787583-ad962015-d1a8-4b58-998f-9589521b8418.png)

the program ramps up memory at `writer.Write()` to 58.5 GB

![image](https://user-images.githubusercontent.com/59346180/188787645-afd48573-00f9-4d5e-81ed-31a4727f221e.png)

and then abruptly ends without warnings/error and without printing 'Finished!' (ends at `writer.Write()`)

![image](https://user-images.githubusercontent.com/59346180/188787788-d91dccbe-78c0-4708-868a-4e879f9ba86c.png)

**Comment by whophil** [comment]
Thanks @adam-grant-hendry. That snippet now fails on my machine after 2 minutes or so of trying. Somehow I was able to write a 3.5 GB file yesterday, but I don't know what exactly is different now.

If I tell the `vtkDataSetWriter` to write to file instead of string, the file it produces is 28 GB. This doesn't seem right.

Finally, if I use the appropriate XML writer, I am able to get past `writer.Write()`
```python
    from vtkmodules.vtkIOXML import vtkXMLImageDataWriter
    writer = vtkXMLImageDataWriter()
    writer.SetInputDataObject(volume)
    writer.SetDataModeToBinary()
    writer.SetCompressionLevel(0)
    writer.SetWriteToOutputString(True)
    writer.Write()
    to_serialize = writer.GetOutputString()
```

I wonder whether the return of `GetOutputString()` will be useful, or if the VTK XML writers will need to modified to implement `GetOutputStdString`.

**Comment by banesullivan** [comment]
I wonder if we should not use `vtkDataSetWriter` and stick with appropriate XML writer (meaning we'd need to implement `__getstate__` for each datatype

**Comment by whophil** [comment]
I guess mentions in discussions don't get cross-referenced. Some benchmarking of pickling with XML writers here:
https://github.com/pyvista/pyvista/discussions/3279

**Comment by whophil** [comment]
This is closed with #3286

**Comment by banesullivan** [comment]
@whophil, do you not have the power to close issues? Because you ought to have that
